candlex/lib/candlex/backend.ex

```diff
@@ -0,0 +1,423 @@
+    @impl true
+    def unquote(op)(out, t) do
+      # TODO: Implement in candle
+      out
+      |> Nx.BinaryBackend.unquote(op)(backend_transfer(t, Nx.BinaryBackend, []))
+      |> backend_transfer(__MODULE__, [])
+    end
+  end
+
+  # Indexed
+
+  @impl true
+  def slice(
+        %T{shape: _output_shape} = out,
+        %T{shape: input_shape} = t,
+        starts,
+        lengths,
+        _strides
+      ) do
+    t
+    |> from_nx()
+    |> narrow(starts, lengths, 0, input_shape)
+    # TODO: Support strides
+    # |> stride(output_shape, lengths, strides)
+    |> to_nx(out)
+  end
+
+  # N-dim
+
+  @impl true
+  def concatenate(%T{} = out, tensors, axis) do
+    tensors
+    |> maybe_upcast()
+    |> Enum.map(&from_nx/1)
+    |> Native.concatenate(axis)
+    |> unwrap!()
+    |> to_nx(out)
+  end
+
+  @impl true
+  def dot(
+        %T{type: _out_type} = out,
+        %T{shape: left_shape, type: _left_type} = left,
+        _left_axes,
+        [] = _left_batched_axes,
+        %T{shape: right_shape, type: _right_type} = right,
+        _right_axes,
+        [] = _right_batched_axes
+      ) when tuple_size(left_shape) == 2 and tuple_size(right_shape) == 2 do
+    Native.matmul(
+      from_nx(left),
+      from_nx(right)
+    )
+    |> unwrap!()
+    |> to_nx(out)
+  end
+
+  # Shape
+
+  @impl true
+  def broadcast(out, %T{} = t, shape, axes) do
+    t
+    |> maybe_reshape(shape, axes)
+    |> from_nx()
+    |> Native.broadcast_to(shape)
+    |> unwrap!()
+    |> to_nx(out)
+  end
+
+  @impl true
+  def reshape(%T{shape: shape} = out, %T{} = t) do
+    from_nx(t)
+    |> Native.reshape(shape)
+    |> unwrap!()
+    |> to_nx(out)
+  end
+
+  @impl true
+  def squeeze(%T{} = out, %T{} = t, axes) do
+    # sort the axes desc so we don't have to decrease the axis numbers after each squeeze
+    for axis <- Enum.sort(axes, :desc), reduce: from_nx(t) do
+      ref ->
+      ref
+      |> Native.squeeze(axis)
+      |> unwrap!()
+    end
+    |> to_nx(out)
+  end
+
+  # Type
+
+  @impl true
+  def as_type(%T{type: type} = out, %T{} = t) do
+    from_nx(t)
+    |> Native.to_type(to_candle_dtype(type))
+    |> unwrap!()
+    |> to_nx(out)
+  end
+
+  @impl true
+  def bitcast(out, tensor) do
+    out
+    |> from_binary(to_binary(tensor), [])
+  end
+
+  # Inspect
+
+  @impl true
+  def inspect(%T{} = tensor, inspect_opts) do
+    limit = if inspect_opts.limit == :infinity, do: :infinity, else: inspect_opts.limit + 1
+
+    tensor
+    |> to_binary(min(limit, Nx.size(tensor)))
+    |> then(&Nx.Backend.inspect(tensor, &1, inspect_opts))
+    |> maybe_add_signature(tensor)
+  end
+
+  defp maybe_add_signature(result, %T{data: %CB{resource: ref}}) when is_reference(ref) do
+    Inspect.Algebra.concat([
+      "Candlex.Backend(#{:erlang.ref_to_list(ref)})",
+      Inspect.Algebra.line(),
+      result
+    ])
+  end
+
+  # defp device_option(_backend_options) do
+  #   # TODO: Support CUDA
+  #   :cpu
+  # end
```

Comment:
My suggestion: add device support sooner than later, because you need to know the device on binary operations and potentially transfer across devices if candle does not do it automatically.

Reply:
Added support for CUDA.

All current tests in `test/candlex_test.exs` are both green on cpu and cuda for me :heavy_check_mark: